src/lib.rs

```diff
@@ -308,165 +173,288 @@ impl KafkaJsonToDelta {
-                            .write_record_batch(&partition_cols, &record_batch)
-                            .await?;
+                        // Since we are writing a record batch to the in-memory parquet file, value buffers must be marked
+                        // dirty so we know we have to drop all buffers and re-seek if a rebalance occurs before the next file write.
+                        value_buffers.mark_dirty();
 
-                        // When the memory buffer meets min bytes per file or allowed latency is met, complete the file and start a new one.
-                        let should_complete_file = delta_writer.buffer_len()
-                            >= self.min_bytes_per_file
-                            || latency_timer.elapsed().as_secs() >= self.allowed_latency;
+                        delta_writer.write_record_batch(&record_batch).await?;
 
-                        debug!(
-                            "Should complete file? {}. delta_writer.buffer_len: {}. timer.elapsed: {}",
-                            should_complete_file,
-                            delta_writer.buffer_len(),
-                            latency_timer.elapsed().as_secs()
+                        info!(
+                            "Record batch written. Current latency is {} milliseconds",
+                            latency_timer.elapsed().as_millis()
                         );
 
-                        if should_complete_file {
-                            info!("Completing parquet file write.");
+                        // Finalize file and write Delta transaction
+                        if self.should_complete_file(&delta_writer, &latency_timer) {
+                            // Reset the latency timer to track allowed latency for the next file
+                            latency_timer = Instant::now();
 
-                            let add = delta_writer.complete_file(&partition_cols).await?;
-                            let action = deltalake::action::Action::add(add);
+                            info!("Writing parquet file.");
+
+                            // Lookup the last transaction version from the Delta Log
+                            let last_txn_version = delta_writer
+                                .last_transaction_version(self.app_id.as_str())
+                                .await;
+
+                            // Create the new WAL entry
+                            let wal_entry = if let Ok(Some(txn_version)) = last_txn_version {
+                                info!(
+                                    "Delta table last txn version for {} is {}.",
+                                    self.app_id, txn_version
+                                );
+                                let last_wal_entry =
+                                    wal.get_entry_by_transaction_id(txn_version).await?;
+                                last_wal_entry.prepare_next(&partition_offsets)
+                            } else {
```

Comment:
Because the `if` statement above is considering `Ok(Some())` I am not sure I agree with the assumption this code is making: the lack of an Ok(Some) meaning that there's no transaction ID and that it's appropriate to start from 1.

Any error returned from `last_transaction_version` would cause the `app_id` to receive a new transaction ID in the write-ahead-log of 1, rather than what it may have already had, right?

Reply:
As mentioned above - an `Ok` means we read the delta log successfully, but there was not a txn entry yet. An error in reading the transaction log will produce an `Err` instead. `Ok(None)` basically means this is the first txn entry.

Reply:
Ah - and now I've spotted the absence of handling Ok(None) vs Err. Will fix.